src/metrics_regression.rs

```diff
@@ -104,13 +113,98 @@ impl<
 
         A::one()
             - (diff.mapv(|x| x * x).sum() - mean_error)
-                / (self.mapv(|x| (x - mean) * (x - mean)).sum() + A::from(1e-10).unwrap())
+                / (compare_to.mapv(|x| (x - mean) * (x - mean)).sum() + A::from(1e-10).unwrap())
+    }
+}
+
+impl<
+        'a,
+        A: 'a + NdFloat + FromPrimitive,
+        D: Data<Elem = A>,
+        T: IntoNdProducer<Item = &'a A, Dim = Ix2, Output = ArrayView<'a, A, Ix2>>,
+    > Regression<'a, A, Ix2, T> for ArrayBase<D, Ix1>
+{
+    fn max_error(&self, compare_to: T) -> A {
+        let compare_to: ArrayView2<'a, A> = compare_to.into_producer();
+        if compare_to.len_of(Axis(1)) > 1 {
+            panic!("Expected single targets array");
```

Comment:
I think that a panic for multi-target comparison is unexpected and should be avoided here. I somehow have the feeling that the whole `metrics_regression` mod is unnecessary complex. Perhaps one way to simplify everything would use associated types like this:
```rust
  pub trait SingleTargetRegression<F: Float>: AsTargets<Elem = F> {
      type Tar: AsTargets<Elem = F>;
  
      fn mean_absolute_error(&self, other: &Self::Tar) -> Result<F> {
          self.try_single_target()?
              .sub(&other.try_single_target()?)
              .mapv(|x| x.abs())
              .mean()
              .ok_or(Error::NotEnoughSamples)
      }
  
      // .. provide all implementations here                                                                                                                                                 
  }
  
  impl<F: Float, D: Data<Elem = F>> SingleTargetRegression<F> for ArrayBase<D, Ix1> {
      type Tar = ArrayBase<D, Ix1>;
  }
  
  impl<F: Float, D: Data<Elem = F>> SingleTargetRegression<F> for ArrayBase<D, Ix2> {
      type Tar = ArrayBase<D, Ix1>;
  }
```
we can then define for which types a single score is expected and introduce a second trait for multi target scoring:
```rust
  pub trait MultiTargetRegression<F: Float>: AsTargets<Elem = F> {
      type Tar: AsTargets<Elem = F>;
  
      fn mean_absolute_error(&self, other: &Self::Tar) -> Result<Array1<F>> {
          self.as_multi_targets()
              .axis_iter(Axis(0))
              .zip(other.as_multi_targets().axis_iter(Axis(0)))
              .map(|(a, b)| a.mean_absolute_error(&b))
              .collect()
      }
  }
  
  impl<F: Float, D: Data<Elem = F>> MultiTargetRegression<F> for ArrayBase<D, Ix2> {
      type Tar = ArrayBase<D, Ix2>;
  }
```
but I can add this in a later PR too, if it's too much work now

Reply:
Traits `SingleTargetRegression` and `MultiTargetRegression` have been implemented and now regression metrics can be computed between:
* One-dimensional array - One-dimensional array
* One-dimensional array - bi-dimensional array
* One-dimensional array - dataset
* bi-dimensional array - bi-dimensional array
* bi-dimensional array - dataset
* dataset - dataset
* dataset - one-dimensional array
* dataset - bi-dimensional array

I also added a new Error variant, but I don't know if that was what you had in mind